### PR TITLE
Better Error Message in `ChainedScheduler` and `SequentialLR`

### LINKED ISSUE #121577: Better error message and type checking in `ChainedScheduler` and `SequentialLR`
### 🚀 The feature, motivation and pitch

Currently in `ChainedScheduler`, it does not check whether all given inputs is a subclass of `LRScheduler`. It should raise a `TypeError` if  there's an non-scheduler type in the given sequence.

### Alternatives

Add error message for TypeError if the given input is not a sequence of LRSchdeuler.
Like the following

```python
class ChainedScheduler(LRScheduler):
    def __init__(self, schedulers: Sequence[LRScheduler]):
        if len(schedulers) < 1:
            raise ValueError("ChainedScheduler expects at least one scheduler to be chained, but got no schedulers.")

        first_scheduler: LRScheduler = None
        for scheduler in schedulers:
            if not isinstance(scheduler, LRScheduler):
                raise TypeError(
                    f"ChainedScheduler expects all schedulers to be of type LRScheduler, "
                    f"but got {type(scheduler)}"
                )
            if first_scheduler is None:
                first_scheduler = scheduler

            else:
                if first_scheduler.optimizer != scheduler.optimizer:
                    raise ValueError(
                        "ChainedScheduler expects all schedulers to belong to the same optimizer, but "
                        f"got schedulers at index {0} and {scheduler} to be different"
                    )
        self._schedulers = schedulers
        self.optimizer = first_scheduler.optimizer
```

### Additional context

Additional modification is to align its inputs with `SequentialLR`, as it has an additional `optimizer` input parameter. The the fix would probably be like 

```python
class ChainedScheduler(LRScheduler):
    def __init__(self, schedulers: Sequence[LRScheduler], optimizer: Optional[Optimizer] = None):
        if len(schedulers) < 1:
            raise ValueError("ChainedScheduler expects at least one schedulers to be chained, but got no schedulers.")

        optimizer = optimizer or schedulers[0].optimizer
        for sch_idx, scheduler in enumerate(schedulers):
            if not isinstance(scheduler, LRScheduler):
                raise TypeError(
                    f"ChainedScheduler expects all schedulers to be of type LRScheduler, "
                    f"but got {type(scheduler)}"
                )

            if not sch_idx and optimizer is None:
                optimizer = scheduler.optimizer

            if optimizer != scheduler.optimizer:
                raise ValueError(
                    "ChainedScheduler expects all schedulers to belong to the same optimizer, but "
                    f"got scheduler {scheduler.__class__.__name__} at index {sch_idx} has {scheduler.optimizer}, "
                    f"which is different from {optimizer.__class__.__name__}."
                )
        self._schedulers = schedulers
        self.optimizer = optimizer
```

_No response_

### PR DESCRIPTION
**Pull request by GdoongMathew**
Fixes #121577

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -667,18 +669,21 @@ class SequentialLR(LRScheduler):
...
+        if len(schedulers) < 1:
+            raise ValueError("Sequential Schedulers expects at least one scheduler, but got no scheduler.")
+
+        for scheduler_idx, scheduler in enumerate(schedulers):
+            if not isinstance(scheduler, LRScheduler):
+                raise TypeError(
+                    f"{self.__class__.__name__} expects all schedulers to be of type LRScheduler, "
+                    f"but got {type(scheduler)}"
+                )
+            if scheduler.optimizer != optimizer:
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:681]: This is also BC breaking. Previously, no optimizer really needs to be passed in (it could be kept as None), but now the optimizer needs to be passed into SequentialLR. We'd probably want to relax this check to allow for more flexibility

**Review by janeyx99 — COMMENTED** [review]
Thanks for opening the PR to clean up the LRScheduler code! I have commented some concerns about the BC implications of this change with the additional error checks.

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -667,18 +669,21 @@ class SequentialLR(LRScheduler):
     """
 
     def __init__(self, optimizer, schedulers, milestones, last_epoch=-1, verbose="deprecated"):
-        for scheduler_idx in range(len(schedulers)):
-            if schedulers[scheduler_idx].optimizer != optimizer:
+        if len(schedulers) < 1:
+            raise ValueError("Sequential Schedulers expects at least one scheduler, but got no scheduler.")
+
+        for scheduler_idx, scheduler in enumerate(schedulers):
+            if not isinstance(scheduler, LRScheduler):
```
[the remark this answers is not in the document]
Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:676]: Hi @janeyx99 , thanks for pointing this out. I do agree that accepting `ReduceLROnPlateau` in `SequentialLR` or `ChainedScheduler` would be weird. However, before and after this change actually nither prevent users from passing ReduceLR in. Do you think if we explicitly raise ValueError if get `ReduceLROnPlateau` in both of the two classes be better?

Thanks~

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:681]: Hi @janeyx99, I think the logic here is still the same as the original [L671](https://github.com/pytorch/pytorch/blob/main/torch/optim/lr_scheduler.py#L671). I do agree on unify the logic between `SequentialLR` and `ChainedScheduler` 's `__init__`, to set the input `optimizer` as optional, however, this would probably break others code bases.

File: torch/optim/lr_scheduler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (janeyx99) [reply torch/optim/lr_scheduler.py:676]: Ah, my point is moreso that since there was precedence where ReduceLROnPlateau was not a strict subclass, others may write their own LRSchedulers without it being a subclass, and if they're trying to plug those LRSchedulers into SequentialLR, this change would be BC breaking for that use case.

If they have this change, the concern about ReduceLROnPlateau changes since that is now a subclass and won't trigger this error.

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -896,22 +902,36 @@ class ChainedScheduler(LRScheduler):
...
         >>>     validate(...)
         >>>     scheduler.step()
     """
 
-    def __init__(self, schedulers):
-        for scheduler_idx in range(1, len(schedulers)):
-            if (schedulers[scheduler_idx].optimizer != schedulers[0].optimizer):
+    def __init__(self, schedulers: Sequence[LRScheduler], optimizer: Optional[Optimizer] = None):
+        if len(schedulers) < 1:
+            raise ValueError("ChainedScheduler expects at least one schedulers to be chained, but got no scheduler.")
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:914]:
```suggestion
            raise ValueError("ChainedScheduler expects at least one scheduler to be chained, but got no scheduler.")
```

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -896,22 +902,36 @@ class ChainedScheduler(LRScheduler):
...
+
+        optimizer = optimizer or schedulers[0].optimizer
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:916]: ah TIL you can or with None

File: torch/optim/lr_scheduler.py [hunk 5]
```diff
@@ -896,22 +902,36 @@ class ChainedScheduler(LRScheduler):
...
+        for sch_idx, scheduler in enumerate(schedulers):
+            if not isinstance(scheduler, LRScheduler):
+                raise TypeError(
+                    f"{self.__class__.__name__} expects all schedulers to be of type LRScheduler, "
+                    f"but got {type(scheduler)}"
+                )
+
+            if not sch_idx and optimizer is None:
+                optimizer = scheduler.optimizer
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:925]: Was this use case allowed before? Could people have passed in an optimizer to just one of the LRSchedulers in the chain and have the rest be empty? If not, this check would be unnecessary.

**Review by janeyx99 — APPROVED** [review]
Looks good overall, left some more remarks

File: torch/optim/lr_scheduler.py (hunk 5 above)
Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:925]: Oh I see. this line is totally unnecessary as its already covered by L916. will remove this.

**Comment by GdoongMathew** [comment]
Hi @janeyx99 
Sorry for the late reply, it took me a while to realize that I have to submit a review for my `Pending` comments a few days ago to be actually posted.

File: torch/optim/lr_scheduler.py [hunk 6]
```diff
@@ -667,18 +669,25 @@ class SequentialLR(LRScheduler):
...
-            if schedulers[scheduler_idx].optimizer != optimizer:
+        if len(schedulers) < 1:
+            raise ValueError(f"{self.__class__.__name__} expects at least one scheduler, but got no scheduler.")
+
+        for scheduler_idx, scheduler in enumerate(schedulers):
+            if not hasattr(scheduler, 'optimizer'):
+                raise TypeError(
+                    f"{self.__class__.__name__} at index {scheduler_idx} should have `optimizer` as its attribute."
+                )
+            if isinstance(scheduler, ReduceLROnPlateau):
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:680]: no, this check is worse than before, because we do support ReduceLROnPlateau now! The other one would be better, but I’m still not sure why the check needs to exist at all.

Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:680]: Hi @janeyx99 
friendly ping. Just to let you know that I'm happy to provide any additional information or clarification if needed, and really thanks for your time for reviewing this PR.

Sincerely.

Comment (janeyx99) [reply torch/optim/lr_scheduler.py:680]: Ah yikes, sorry this fell under my stack of notifications--don't be shy about pinging if I am not responding--I don't intend to ignore these!

I see your intention is to treat ReduceLROnPlateau differently due to it needing kwargs in step (and thus being different from other LRSchedulers), yes? If so, why is the check in SequentialLR and not also in Chained below? Are there any cases where having a list of schedulers including ReduceLROnPlateau is fine but would be invalidated by this check? If there are no use cases where people can use Chained/Sequential with ReduceLROnPlateau, then this check is fine, though it may be good to specify that the reason why is due to ReduceLROnPlateau requiring kwargs in step to be specified.

Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:680]: Thank you~
To answer your question, yes, it indeed is the reason why I did this check, and thanks for pointing this out, this check should be applied to both Chain and Sequential scheduler. 
Should I then change the error message to the following and add it to both Chain and Sequential?
```python
if isinstance(scheduler, ReduceLROnPlateau):
        raise ValueError(
            f"{self.__class__.__name__} does not support `ReduceLROnPlateau` scheduler as it requires additional kwargs to be specified when calling `step`, but got one at index {scheduler_idx} in the given schedulers sequence."
        )
```

Comment (janeyx99) [reply torch/optim/lr_scheduler.py:680]: Yes

Comment (GdoongMathew) [reply torch/optim/lr_scheduler.py:680]: Fixed.

**Comment by GdoongMathew** [comment]
solve merge conflicts.

**Comment by janeyx99** [comment]
Do you have lintrunner setup locally? it's p easy to set up!

`pip install lintrunner`
`lintrunner init`
`lintrunner -a`

This way we won't get blocked like this all the time (which is esp time consuming cuz i have to remember to approve the workflows 😅)

**Comment by GdoongMathew** [comment]
Thanks for the hint, will fix this later.

**Comment by GdoongMathew** [comment]
@janeyx99 linter error should be fixed, thanks~
